File: examples/js/producer.js
const net = require('net');

function produce(host, port, topic, key, value) {
    const client = new net.Socket();
    client.connect(port, host, () => {
        const topicBuf = Buffer.from(topic, 'utf-8');
        const keyBuf = Buffer.from(key);
        const valBuf = Buffer.from(value);

        // Calculate request body size:
        // 1 byte request type + 2 bytes topic len + topic + 4 bytes key len + key + 4 bytes val len + val
        const bodySize = 1 + 2 + topicBuf.length + 4 + keyBuf.length + 4 + valBuf.length;
        const frame = Buffer.alloc(4 + bodySize);

        let offset = 0;
        frame.writeUInt32BE(bodySize, offset);
        offset += 4;
        frame.writeUInt8(1, offset); // ReqProduce = 1
        offset += 1;
        frame.writeUInt16BE(topicBuf.length, offset);
        offset += 2;
        topicBuf.copy(frame, offset);
        offset += topicBuf.length;
        frame.writeUInt32BE(keyBuf.length, offset);
        offset += 4;
        keyBuf.copy(frame, offset);
        offset += keyBuf.length;
        frame.writeUInt32BE(valBuf.length, offset);
        offset += 4;
        valBuf.copy(frame, offset);

        client.write(frame);
    });

    client.on('data', (data) => {
        if (data.length < 4) return;
        const respLen = data.readUInt32BE(0);
        const body = data.subarray(4, 4 + respLen);
        const status = body.readUInt8(0);
        if (status === 0) {
            const offsetVal = body.readBigInt64BE(1);
            console.log(`Produced! Status: OK, Offset: ${offsetVal}`);
        } else {
            console.log("Produce failed");
        }
        client.destroy();
    });
}

produce('localhost', 9092, 'test', 'jskey', 'jsval');

File: examples/python/producer.py
import socket
import struct

def produce(addr, topic, key, value):
    s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    s.connect(addr)

    # String fields: 2-byte len + data
    topic_bytes = topic.encode('utf-8')
    topic_prefix = struct.pack('>H', len(topic_bytes))

    # Key / Value fields: 4-byte len + data
    key_prefix = struct.pack('>I', len(key))
    val_prefix = struct.pack('>I', len(value))

    # Combine body payload: ReqProduce = 1
    body = b'\x01' + topic_prefix + topic_bytes + key_prefix + key + val_prefix + value

    # 4-byte frame length prefix
    header = struct.pack('>I', len(body))
    s.sendall(header + body)

    # Read response: 4-byte length prefix
    resp_len_buf = s.recv(4)
    if len(resp_len_buf) < 4:
        print("Failed to read header")
        return
    resp_len = struct.unpack('>I', resp_len_buf)[0]
    resp_body = s.recv(resp_len)

    # Response starts with status (1-byte, OK = 0)
    status = resp_body[0]
    if status == 0:
        # Next is logical offset (8-byte int64)
        offset = struct.unpack('>q', resp_body[1:9])[0]
        print(f"Produced! Status: OK, Offset: {offset}")
    else:
        print("Produced failed with error status")
    s.close()

if __name__ == "__main__":
    produce(('localhost', 9092), 'test', b'pykey', b'pyval')
